### PR TITLE
Add load limit

### PR DESCRIPTION
**Pull request by yoavkatz**
Added a env new variable called UNITEXT_IBM_COS_LOADER_LINE_LIMIT which is used to limit the number of lines downloaded from COS.  This is used to speedup generating datacards (using the prepare/cards/*.py files) .  

This works only for datsets that are in jsonl format.   

In another change avoid placing random file names in output
(To simplify diffs comparison)

### CODE REVIEW

**Review by OfirArviv — COMMENTED** [review]
what is the use case? In general I'm not in favor on environment params unless there is a good reason

**Comment by yoavkatz** [comment]
> what is the use case? In general I'm not in favor on environment params unless there is a good reason

The use case is that if I want to 'prepare' json cards from all the .py files , I don't really need to load all the data (can be 100MB) from COS.  This can reduce the time considerably - saving a minute for each card generated.

File: src/unitxt/loaders.py [hunk 1]
```diff
@@ -19,6 +20,13 @@
 
 
 class Loader(SourceOperator):
+    # The loader_limit an optional parameter used to control the maximum number of instances to load from the the source.
+    # It is usually provided to the loader via the recipe (see standard.py)
+    # The loader can use this value to limit the amount of data downloaded from the source
+    # to reduce loading time.  However, this may not always be possible, so the
+    # loader may ingore this.  In any case, the recipe, will limit the number of instances in the returned
```
Comment (OfirArviv) [inline src/unitxt/loaders.py:27]: is the loader per card?
if so, isn't it possible to error out if its not possible to limit the load?
because the source is fixed in the card
Why not to throw an error?

File: src/unitxt/loaders.py [hunk 2]
```diff
@@ -62,13 +70,24 @@ class LoadFromIBMCloud(Loader):
...
-        print(f"Downloading {item_name} from {bucket_name} COS to {local_file}")
+        print(f"Downloading {item_name} from {bucket_name} COS")
         try:
             response = cos.Object(bucket_name, item_name).get()
             size = response["ContentLength"]
+            body = response["Body"]
         except Exception as e:
             raise Exception(f"Unabled to access {item_name} in {bucket_name} in COS", e)
 
+        if self.loader_limit != None:
```
Comment (OfirArviv) [inline src/unitxt/loaders.py:81]: you are still downloading everything, no? What am I missing?

File: src/unitxt/operators.py [hunk 3]
```diff
@@ -858,6 +858,9 @@ def process(self, stream: Stream, stream_name: str = None) -> Generator:
         for instance in stream:
             counter[self.signature(instance)] += 1
 
+        if len(counter) == 0:
```
Comment (OfirArviv) [inline src/unitxt/operators.py:861]: why can we get 0? should we allow that?

File: src/unitxt/standard.py [hunk 4]
```diff
@@ -132,10 +158,21 @@ class StandardRecipeWithIndexes(BaseRecipe):
     def prepare(self):
         assert (
             self.template_card_index is None or self.template is None
```
Comment (OfirArviv) [inline src/unitxt/standard.py:160]: is this related  another issue?

File: src/unitxt/standard.py [hunk 5]
```diff
@@ -132,10 +158,21 @@ class StandardRecipeWithIndexes(BaseRecipe):
...
-        ), "Specify either template or template_card_index"
+        ), f"Specify either template ({self.template}) or template_card_index ({self.template_card_index}) but not both"
+        assert (
+            not self.template_card_index is None or not self.template is None
```
Comment (OfirArviv) [inline src/unitxt/standard.py:163]: confusing  logic

File: src/unitxt/standard.py [hunk 6]
```diff
@@ -50,12 +52,36 @@ def verify(self):
                 raise ValueError(
                     f"demos_pool_size must be bigger than num_demos={self.num_demos}, Got demos_pool_size={self.demos_pool_size}"
                 )
+            if self.loader_limit and self.demos_pool_size > self.loader_limit:
```
Comment (OfirArviv) [inline src/unitxt/standard.py:55]: you have if self.loader_limit in all of them. Do a single if on top of it all.

Also what if train_validatation > the limit for example?

File: src/unitxt/loaders.py (hunk 2 above)
Comment (yoavkatz) [reply src/unitxt/loaders.py:81]: Loader limit is set to each of the loaders by the calling recipe.

File: src/unitxt/operators.py (hunk 3 above)
Comment (yoavkatz) [reply src/unitxt/operators.py:861]: It can happen if by mistake someone filters too many results.  In most cases, it should not happen, once I added the checks in the standard recipe verify method, but it can happen .    The question is more general - can streams be empty.   But we we allow empty streams, we need operators to behave correctly over them.

File: src/unitxt/standard.py (hunk 6 above)
Comment (yoavkatz) [reply src/unitxt/standard.py:55]: I'll move the if to the top.

self.max_validation_instances is checked below.

File: requirements/tests.rqr [hunk 7]
```diff
@@ -1,3 +1,4 @@
 bert_score
 transformers
 sentence_transformers
+ibm-cos-sdk
```
Comment (OfirArviv) [inline requirements/tests.rqr:4]: we test ibm-cos here? how?

File: src/unitxt/standard.py [hunk 8]
```diff
@@ -50,12 +52,35 @@ def verify(self):
...
+                raise ValueError(
+                    f"demos_pool_size must be bigger than loader_limit{self.loader_limit}, Got demos_pool_size={self.demos_pool_size}"
+                )
```
Comment (OfirArviv) [inline src/unitxt/standard.py:58]: this one is not in the general if

File: src/unitxt/test_utils/card.py [hunk 9]
```diff
@@ -26,44 +27,38 @@ def test_metrics_exist(card):
...
         assert json.dumps(card_.to_dict(), sort_keys=True) == json.dumps(
             card.to_dict(), sort_keys=True
         ), "Card loaded is not equal to card stored"
 
 
-def load_examples_from_common_recipe(card, tested_split):
-    if card.templates:
-        num_templates = len(card.templates)
+def load_examples_from_standard_recipe(card, tested_split, template_card_index):
+    print("=" * 80)
```
Comment (OfirArviv) [inline src/unitxt/test_utils/card.py:39]: is this intentional?

File: src/unitxt/test_utils/card.py [hunk 10]
```diff
@@ -26,44 +27,38 @@ def test_metrics_exist(card):
...
-    recipe = CommonRecipe(
+    recipe = StandardRecipe(
         card=card,
         demos_pool_size=100,
         demos_taken_from=tested_split,
         num_demos=3,
-        template_item=template_item,
-        instruction_item=0 if num_instructions else None,
+        template_card_index=template_card_index,
+        instruction_card_index=0 if num_instructions else None,
```
Comment (OfirArviv) [inline src/unitxt/test_utils/card.py:49]: this is never None according to line 48

File: tests/test_recipe.py [hunk 11]
```diff
@@ -112,6 +112,77 @@ def test_standard_recipe_with_balancer(self):
...
+            instruction="instructions.models.llama",
+            template="templates.key_val",
+            format="formats.user_agent",
+            demos_pool_size=5,
+            num_demos=1,
+            loader_limit=10,
+        )
+
+        stream = recipe()
+        self.assertEqual(len(list(stream["train"])), 5)
```
Comment (OfirArviv) [inline tests/test_recipe.py:127]: this is confusing, why only 5 and not 10 also?

File: tests/test_recipe.py [hunk 12]
```diff
@@ -112,6 +112,77 @@ def test_standard_recipe_with_balancer(self):
...
+    def test_standard_recipe_with_template_errors(self):
+        with self.assertRaises(AssertionError) as cm:
+            recipe = StandardRecipeWithIndexes(card="cards.wnli")
+        self.assertEqual(str(cm.exception), "Specify either template or template_card_index in card")
+
+        with self.assertRaises(AssertionError) as cm:
+            recipe = StandardRecipeWithIndexes(
+                card="cards.wnli", template="templates.key_val", template_card_index=100
+            )
+        self.assertTrue("Specify either template" in str(cm.exception))
```
Comment (OfirArviv) [inline tests/test_recipe.py:172]: you are missing " or template_card_index in card" I think , and the assert is weird, why 2?

Comment (OfirArviv) [reply tests/test_recipe.py:172]: maybe a better idea is to add a short description to each test, to explain the scenario

File: requirements/tests.rqr (hunk 7 above)
Comment (yoavkatz) [reply requirements/tests.rqr:4]: We test the API using mocks. See test_loaders.py.

File: src/unitxt/standard.py (hunk 8 above)
Comment (yoavkatz) [reply src/unitxt/standard.py:58]: Right, because it also requires multi  conditions on the poll size number of demos, which are in the inclosing if statements.

File: src/unitxt/test_utils/card.py (hunk 9 above)
Comment (yoavkatz) [reply src/unitxt/test_utils/card.py:39]: Yes. It makes the output more readible.

File: src/unitxt/test_utils/card.py (hunk 10 above)
Comment (yoavkatz) [reply src/unitxt/test_utils/card.py:49]: num_instructions is never None, but 
  
  "if num_instructions " does not check if the (int) variable is None, it checks if it's non zero:
  
  https://stackoverflow.com/questions/73241173/what-is-the-meaning-of-the-if-0-in-python

File: tests/test_recipe.py (hunk 11 above)
Comment (yoavkatz) [reply tests/test_recipe.py:127]: Because 5 elements are taken to the demo pool.  Added comment.

File: tests/test_recipe.py (hunk 12 above)
Comment (yoavkatz) [reply tests/test_recipe.py:172]: Changed to regex search and added short description

File: src/unitxt/catalog.py [hunk 13]
```diff
@@ -97,16 +100,23 @@ def __contains__(self, artifact_identifier: str):
 def verify_legal_catalog_name(name):
     assert re.match(
         r"^[\w" + COLLECTION_SEPARATOR + "]+$", name
-    ), 'Catalog name should be alphanumeric, ":" should specify dirs (instead of "/").'
+    ), f'Catalog name should be alphanumeric, ":" should specify dirs (instead of "/"). Name: {name}'
```
Comment (OfirArviv) [inline src/unitxt/catalog.py:103]: I think this error is phrased in a confusing way, but it is not a blocker

File: src/unitxt/standard.py [hunk 14]
```diff
@@ -48,14 +50,37 @@ def verify(self):
                 )
             if self.demos_pool_size < self.num_demos:
                 raise ValueError(
-                    f"demos_pool_size must be bigger than num_demos={self.num_demos}, Got demos_pool_size={self.demos_pool_size}"
+                    f"demos_pool_size must be bigger than num_demos ({self.num_demos}), Got demos_pool_size={self.demos_pool_size}"
+                )
+            if self.loader_limit and self.demos_pool_size > self.loader_limit:
```
Comment (OfirArviv) [inline src/unitxt/standard.py:55]: I thin it should be >=. We don't want all the examples to go into the demo pool either, right?

**Review by OfirArviv — APPROVED** [review]

File: src/unitxt/catalog.py (hunk 13 above)
Comment (yoavkatz) [reply src/unitxt/catalog.py:103]: Improved

File: src/unitxt/standard.py (hunk 14 above)
Comment (yoavkatz) [reply src/unitxt/standard.py:55]: Not clear - if you don't do fine/prompt tuning it's not a problem to take all training to the demo pool.
